# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/frontend/ForInEmitter.cpp
# Commit: 7a3a4207c55b
# Full Hash: 7a3a4207c55b27d6a73783c0021cfed2368bdbb8
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2019-12-06 21:48:33
# Regressor Bug: 1601897
# File Overlap Count: 3
# Description:
#   Bug 1601897 part 2 - Move patchBreaks and addTryNote calls to LoopControl::emitLoopEnd. r=arai
#   
#   This ensures all loops handle this the same way.
#   
#   For JSTRY_FOR_IN some places asserted the try note ends at the
# ==============================================================================

diff -r 7a9ab8423e85 -r 7a3a4207c55b js/src/frontend/ForInEmitter.cpp
--- a/js/src/frontend/ForInEmitter.cpp	Fri Dec 06 11:08:02 2019 +0000
+++ b/js/src/frontend/ForInEmitter.cpp	Fri Dec 06 11:08:04 2019 +0000
@@ -138,7 +138,7 @@
     //              [stack] ITER
     return false;
   }
-  if (!loopInfo_->emitLoopEnd(bce_, JSOP_GOTO)) {
+  if (!loopInfo_->emitLoopEnd(bce_, JSOP_GOTO, JSTRY_FOR_IN)) {
     //              [stack] ITER
     return false;
   }
@@ -149,23 +149,12 @@
   MOZ_ASSERT(stackDepth == loopDepth_);
   bce_->bytecodeSection().setStackDepth(stackDepth);
 
-  if (!loopInfo_->patchBreaks(bce_)) {
-    //              [stack] ITER ITERVAL
-    return false;
-  }
-
   // Pop the enumeration value.
   if (!bce_->emit1(JSOP_POP)) {
     //              [stack] ITER
     return false;
   }
 
-  if (!bce_->addTryNote(JSTRY_FOR_IN, bce_->bytecodeSection().stackDepth(),
-                        loopInfo_->headOffset(),
-                        bce_->bytecodeSection().offset())) {
-    return false;
-  }
-
   if (!bce_->emit1(JSOP_ENDITER)) {
     //              [stack]
     return false;